### PR TITLE
Clean Up FastHotStuff Block Proposal End to End Flow

### PR DESCRIPTION
**Pull request by tholonious**

### CODE REVIEW

**Comment by tholonious** [comment]
Current dependencies on/for this PR:
* feature/proof-of-stake
  * **PR #830** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/830?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #848** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/848?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #849** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/849?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #850** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/850?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #851** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/851?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/851?utm_source=stack-comment).

File: consensus/event_loop.go [hunk 1]
```diff
@@ -782,6 +782,7 @@ func (fc *fastHotStuffEventLoop) tryConstructTimeoutQCInCurrentView() *FastHotSt
 		View:           fc.currentView,                          // The view that the timeout QC is proposed in
 		TipBlockHash:   validatorsHighQC.GetBlockHash(),         // The block hash that we extend from
 		TipBlockHeight: safeBlock.GetHeight(),                   // The block height that we extend from
+		QC:             validatorsHighQC,                        // The high QC aggregated from the timeout messages
```
Comment (lazynina) [inline consensus/event_loop.go:785]: Are we expecting the QC field to be populated for a timeout QC? we already have the high QC in the AggregateQC struct.

File: lib/pos_consensus_controller.go [hunk 2]
```diff
@@ -28,13 +29,50 @@ func NewConsensusController(params *DeSoParams, blockchain *Blockchain, mempool
...
-func (cc *ConsensusController) Init() {
-	// This initializes the FastHotStuffEventLoop based on the blockchain state. This should
-	// only be called once the blockchain has synced, the node is ready to join the validator
-	// network, and the node is able validate blocks in the steady state.
-	//
-	// TODO: Implement this later once the Blockchain struct changes are merged. We need to be
-	// able to fetch the tip block and current persisted view from DB from the Blockchain struct.
+// ConsensusController.Start initializes and starts the FastHotStuffEventLoop based on the
+// blockchain state. This should only be called once the blockchain has synced, the node is
+// ready to join the validator network, and the node is able validate blocks in the steady state.
```
Comment (lazynina) [inline lib/pos_consensus_controller.go:34]:
```suggestion
// ready to join the validator network, and the node is able to validate blocks in the steady state.
```

File: lib/pos_consensus_controller.go [hunk 3]
```diff
@@ -28,13 +29,50 @@ func NewConsensusController(params *DeSoParams, blockchain *Blockchain, mempool
...
+	tipBlock := cc.blockchain.BlockTip()
+
+	// Fetch the validator set at each safe block
+	tipBlockWithValidators, err := cc.fetchValidatorListsForSafeBlocks([]*MsgDeSoHeader{tipBlock.Header})
+	if err != nil {
+		return errors.Errorf("ConsensusController.Start: Error fetching validator list for tip blocks: %v", err)
+	}
+
+	// Fetch the safe blocks that are eligible to be extended from by the next incoming tip block
+	safeBlocks, err := cc.blockchain.GetSafeBlocks()
```
Comment (lazynina) [inline lib/pos_consensus_controller.go:56]: This reminds me that I should add a comment on `GetSafeBlocks` that the caller should acquire a read lock before calling.

File: lib/pos_consensus_controller.go [hunk 4]
```diff
@@ -90,35 +128,16 @@ func (cc *ConsensusController) handleBlockProposerEvent(
...
 ) error {
-	// Validate that the expected event type is a block proposal event type
-	possibleExpectedEventTypes := []consensus.FastHotStuffEventType{
-		consensus.FastHotStuffEventTypeConstructVoteQC,
-		consensus.FastHotStuffEventTypeConstructTimeoutQC,
-	}
-	if !collections.Contains(possibleExpectedEventTypes, expectedEventType) {
-		return errors.Errorf("Invalid expected event type: %v", expectedEventType)
+	// Validate that the event's type is the expected proposal event type
+	if !isValidateBlockProposalEvent(event, expectedEventType) {
```
Comment (lazynina) [inline lib/pos_consensus_controller.go:132]:
```suggestion
	if !isValidBlockProposalEvent(event, expectedEventType) {
```
Nit on naming of the function

File: lib/pos_consensus_controller.go [hunk 5]
```diff
@@ -475,6 +457,66 @@ func (cc *ConsensusController) tryProcessBlockAsNewTip(block *MsgDeSoBlock) ([]*
...
+	// to construct a block based on the UtxoView. In practice, this lock ends up being a no-op but it guarantees
+	// thread-safety by making no assumptions about how other parts of the codebase operate outside of this struct.
+	cc.blockchain.ChainLock.RLock()
+	defer cc.blockchain.ChainLock.RUnlock()
+
+	// Get the parent block's hash
+	parentBlockHash := BlockHashFromConsensusInterface(event.QC.GetBlockHash())
+
+	// Build a UtxoView at the parent block
+	utxoViewAtParent, err := cc.blockchain.getUtxoViewAtBlockHash(*BlockHashFromConsensusInterface(event.QC.GetBlockHash()))
```
Comment (lazynina) [inline lib/pos_consensus_controller.go:477]:
```suggestion
	utxoViewAtParent, err := cc.blockchain.getUtxoViewAtBlockHash(*parentBlockHash)
```

File: lib/pos_consensus_controller.go [hunk 6]
```diff
@@ -568,11 +610,34 @@ func getEpochEntryForBlockHeight(blockHeight uint64, epochEntries []*EpochEntry)
...
 }
 
-func (bav *UtxoView) SimulateNextEpochEntry(epochEntry *EpochEntry) (*EpochEntry, error) {
-	return bav.computeNextEpochEntry(
-		epochEntry.EpochNumber,
-		epochEntry.FinalBlockHeight,
-		epochEntry.InitialView,
-		epochEntry.CreatedAtBlockTimestampNanoSecs,
-	)
+func isValidateBlockProposalEvent(event *consensus.FastHotStuffEvent, expectedEventType consensus.FastHotStuffEventType) bool {
```
Comment (lazynina) [inline lib/pos_consensus_controller.go:613]:
```suggestion
func isValidBlockProposalEvent(event *consensus.FastHotStuffEvent, expectedEventType consensus.FastHotStuffEventType) bool {
```

**Review by lazynina — CHANGES_REQUESTED** [review]

File: consensus/event_loop.go (hunk 1 above)
Comment (tholonious) [reply consensus/event_loop.go:785]: Correct. The QC field is populated in the timeout QC event just for convenience. The reference implementation populates it too for convenience. This way, the QC field just tells us the high QC that we will extend the next block from.

The QC field in the final proposed block won't be populated though

File: lib/pos_consensus_controller.go (hunk 3 above)
Comment (tholonious) [reply lib/pos_consensus_controller.go:56]: Done. I added a comment for it above the GetSafeBlocks function signature.

**Review by lazynina — APPROVED** [review]

**Pull request merged**